internal/core/metadata/operators/device/profile/update.go

```diff
@@ -0,0 +1,141 @@
+	"strconv"
+
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+
+	errors2 "github.com/edgexfoundry/edgex-go/internal/core/metadata/errors"
+)
+
+// UpdateDeviceProfileExecutor provides functionality for updating and persisting device profiles.
+// Returns NewErrDuplicateName
+// Returns ErrDeviceProfileNotFound if a device profile could not be found with a matching ID nor name
+// Returns ErrDeviceProfileInvalidState if the device profile has one or more devices or provision watchers
+// associated with it
+type UpdateDeviceProfileExecutor interface {
+	Execute() (contract.DeviceProfile, error)
+}
+
+// updateDeviceProfile updates device profiles by way of the operator pattern.
+type updateDeviceProfile struct {
+	database DeviceProfileUpdater
+	dp       contract.DeviceProfile
+}
+
+// Execute updates and persists the device profile.
+func (n updateDeviceProfile) Execute() (contract.DeviceProfile, error) {
+	// Check if the Device Profile exists
+	var to contract.DeviceProfile
+	// First try with ID
+	to, err := n.database.GetDeviceProfileById(n.dp.Id)
+	if err != nil {
+		// Try with name
+		to, err = n.database.GetDeviceProfileByName(n.dp.Name)
+		if err != nil {
+			return contract.DeviceProfile{}, errors2.NewErrDeviceProfileNotFound(n.dp.Id, n.dp.Name)
+		}
+	}
+
+	d, err := n.database.GetDevicesByProfileId(to.Id)
+	if err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	if len(d) > 0 {
+		// LoggingClient.Debug(strconv.Itoa(len(d))+"devices are associated with the device profile named: " + to.Name)
+		// http.Error(w, "Conflict", http.StatusConflict)
+		return contract.DeviceProfile{}, errors2.NewErrDeviceProfileInvalidState(n.dp.Id, n.dp.Name, strconv.Itoa(len(d))+" devices are associated with the device profile")
+	}
+
+	p, err := n.database.GetProvisionWatchersByProfileId(to.Id)
+	if err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	if len(p) > 0 {
+
+		return contract.DeviceProfile{}, errors2.NewErrDeviceProfileInvalidState(n.dp.Id, n.dp.Name, strconv.Itoa(len(p))+" provision watchers are associated with the device profile")
+	}
+
+	// Update the device profile fields based on the passed JSON
+	if err := n.updateDeviceProfileFields(n.dp, &to); err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	if err := n.database.UpdateDeviceProfile(to); err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	return to, nil
+}
+
+// NewUpdateDeviceProfileExecutor creates an UpdateDeviceProfileExecutor.
+func NewUpdateDeviceProfileExecutor(db DeviceProfileUpdater, dp contract.DeviceProfile) UpdateDeviceProfileExecutor {
+	return updateDeviceProfile{
+		database: db,
+		dp:       dp}
+}
+
+// Update the fields of the device profile
+// to - the device profile that was already in Mongo (whose fields we're updating)
+// from - the device profile that was passed in with the request
+func (n updateDeviceProfile) updateDeviceProfileFields(from contract.DeviceProfile, to *contract.DeviceProfile) error {
```

Comment:
I know this is a previously defined function and you didn't write it, just moved it into the operator from the router. I _**cannot stand**_ functions that mutate their parameters. I cannot tell you how many bugs I've had to root cause because of code written this way. Suggest this be refactored to return the new DeviceProfile rather than mutate the `*to` parameter. The caller should be changed to expect the new return as well.

Reply:
:+1: Excellent point and I agree